// @appen-china/easy-form/es/components/CascaderField.js
import React, { useCallback } from 'react';
import { Cascader } from 'antd';
import { CaretDownOutlined, CaretUpOutlined } from '@ant-design/icons';
import { LayoutMode } from '../types';

var CascaderField = function CascaderField(_ref) {
  var onChange = _ref.onChange,
      value = _ref.value,
      _ref$options = _ref.options,
      options = _ref$options === void 0 ? [] : _ref$options,
      disabled = _ref.disabled,
      layoutMode = _ref.layoutMode,
      _ref$optionSearchable = _ref.optionSearchable,
      optionSearchable = _ref$optionSearchable === void 0 ? false : _ref$optionSearchable,
      _ref$clickStopPropaga = _ref.clickStopPropagation,
      clickStopPropagation = _ref$clickStopPropaga === void 0 ? false : _ref$clickStopPropaga;

  var filter = function filter(inputValue, path) {
    return path.some(function (option) {
      return option.label.toLowerCase().indexOf(inputValue.toLowerCase()) > -1;
    });
  };

  var preventDefaultClick = useCallback(function (event) {
    if (clickStopPropagation) {
      event.stopPropagation();
      event.preventDefault();
    }
  }, [clickStopPropagation]);
  return /*#__PURE__*/React.createElement(Cascader, {
    allowClear: false,
    defaultValue: value,
    suffixIcon: layoutMode === LayoutMode.COMPACT ? function () {
      return /*#__PURE__*/React.createElement("span", null, /*#__PURE__*/React.createElement(CaretUpOutlined, null), /*#__PURE__*/React.createElement(CaretDownOutlined, null));
    } : null,
    options: options,
    onChange: onChange,
    disabled: disabled,
    expandTrigger: "click",
    placement: layoutMode === LayoutMode.COMPACT ? 'bottomRight' : 'bottomLeft',
    showSearch: optionSearchable ? {
      filter: filter
    } : null,
    dropdownClassName: layoutMode === LayoutMode.COMPACT ? 'easyform-cascader-dropdown-compact' : null,
    dropdownRender: function dropdownRender(menu) {
      return /*#__PURE__*/React.createElement("div", {
        role: "button",
        tabIndex: 0,
        onClick: function onClick() {},
        onKeyDown: preventDefaultClick
      }, menu);
    }
  });
};

export default CascaderField;
