// @appen-china/easy-form/es/types/index.js
export var FieldControlType;

(function (FieldControlType) {
  FieldControlType["TEXT"] = "TEXT";
  FieldControlType["TEXTAREA"] = "TEXTAREA";
  FieldControlType["IMAGETEXT"] = "IMAGETEXT";
  FieldControlType["SWITCH"] = "SWITCH";
  FieldControlType["CHECKBOX"] = "CHECKBOX";
  FieldControlType["RADIO"] = "RADIO";
  FieldControlType["SELECT"] = "SELECT";
  FieldControlType["CASCADER"] = "CASCADER";
})(FieldControlType || (FieldControlType = {}));

export var FieldValueType;

(function (FieldValueType) {
  FieldValueType["STRING"] = "string";
  FieldValueType["NUMBER"] = "number";
  FieldValueType["BOOLEAN"] = "boolean";
})(FieldValueType || (FieldValueType = {}));

export var ConditionType;

(function (ConditionType) {
  ConditionType["FIELD"] = "FIELD";
  ConditionType["AGGREGATION"] = "AGGREGATION";
})(ConditionType || (ConditionType = {}));

export var FieldOperatorType;

(function (FieldOperatorType) {
  FieldOperatorType["EQUAL"] = "equal_to";
  FieldOperatorType["NOT_EQUAL"] = "not_equal_to";
  FieldOperatorType["GREATER"] = "greater_than";
  FieldOperatorType["GREATER_EQUAL"] = "greater_than_or_equal_to";
  FieldOperatorType["LESS"] = "less_than";
  FieldOperatorType["LESS_EQUAL"] = "less_than_or_equal_to";
  FieldOperatorType["CONTAINS"] = "contains";
  FieldOperatorType["START"] = "start_with";
  FieldOperatorType["END"] = "end_with";
})(FieldOperatorType || (FieldOperatorType = {}));

export var AggregationOperatorType;

(function (AggregationOperatorType) {
  AggregationOperatorType["AND"] = "and";
  AggregationOperatorType["OR"] = "or";
})(AggregationOperatorType || (AggregationOperatorType = {}));

export var OptionType;

(function (OptionType) {
  OptionType["BUTTON"] = "button";
  OptionType["DEFAULT"] = "default";
})(OptionType || (OptionType = {}));

export var Mode;

(function (Mode) {
  Mode["MULTIPLE"] = "multiple";
})(Mode || (Mode = {}));

export var LayoutMode;

(function (LayoutMode) {
  LayoutMode["COMPACT"] = "compact";
})(LayoutMode || (LayoutMode = {}));
